lib/improver/nowcasting/optical_flow.py

```diff
@@ -323,21 +323,23 @@ def grid_spacing(coord):
         new_datetime = original_datetime + timestep
         new_time = (cube.coord("time").units).date2num(new_datetime)
 
-        if np.int64(new_time) == new_time:
-            new_time = np.int64(new_time)
-        else:
-            new_time = np.float64(new_time)
-
-        advected_cube.coord("time").points = new_time
+        advected_cube.coord("time").convert_units(
+            "seconds since 1970-01-01 00:00:00")
+        advected_cube.coord("forecast_reference_time").convert_units(
+            "seconds since 1970-01-01 00:00:00")
```

Comment:
I don't think this does what it needs to do.  What you have done here is:

1. Calculate the time point in the units of the original cube (may be hours)
2. Convert the advected cube time and frt coordinate units to seconds
3. Round the time point (in original units) an integer value
4. Assign the time point value (in its original units) to the advected cube

I think you need to do this in a different order.  Since advected_cube is an unmodified copy of the original, you could:

1. Calculate the time point in the units of the original cube
2. Assign that point to the advected cube (which is still in the original units), without changing the datatype (as the point may not yet be an integer)
3. Update the advected cube time coordinate to have the correct units
4. Cast the time coordinate points to int64 (now that we know they are in integer seconds)

Reply:
I think I've now corrected the order of the operations, as you suggest.